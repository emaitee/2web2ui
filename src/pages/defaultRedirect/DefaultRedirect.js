import React, { Component } from 'react';
import { Loading } from 'src/components/loading/Loading';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import selectAccessConditionState from 'src/selectors/accessConditionState';
import config from 'src/config';

export class DefaultRedirect extends Component {

  componentDidMount() {
    this.handleRedirect();
  }

  componentDidUpdate() {
    this.handleRedirect();
  }

  handleRedirect() {
    const { location, history, currentUser, ready } = this.props;
    const { state: routerState = {}} = location;

    // if there is a redirect route set on state, we can
    // redirect there before access condition state is ready
    if (routerState.redirectAfterLogin) {
      history.push(routerState.redirectAfterLogin, { replace: true });
      return;
    }

    // if access condition state hasn't loaded, we can't
    // make a redirect decision yet
    if (!ready) {
      return;
    }

    // reporting users are all sent to the summary report
    if (currentUser.access_level === 'reporting') {
      history.push('/reports/summary', { replace: true });
      return;
    }

    // everyone else is sent to the config.splashPage route
    history.push(config.splashPage, { replace: true });
  }

  render() {
    return <Loading />;
  }
}

const mapStateToProps = (state) => selectAccessConditionState(state);
export default withRouter(connect(mapStateToProps)(DefaultRedirect));
