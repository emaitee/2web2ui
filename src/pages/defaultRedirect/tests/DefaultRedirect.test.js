import React from 'react';
import { shallow } from 'enzyme';
import { DefaultRedirect } from '../DefaultRedirect';

jest.mock('src/config', () => ({
  splashPage: '/config/splash'
}));

describe('Component: DefaultRedirect', () => {

  let props;
  let wrapper;
  let instance;

  beforeEach(() => {
    props = {
      location: {
        state: {}
      },
      history: {
        push: jest.fn()
      },
      currentUser: {
        access_level: 'admin'
      },
      ready: false
    };
    props.history.push.mockName('historyPushMock');
    wrapper = shallow(<DefaultRedirect {...props} />);
    instance = wrapper.instance();
  });

  it('should render correctly by default', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect after login', () => {
    wrapper.setProps({ location: { state: { redirectAfterLogin: '/test/redirect/after/login' }}});
    props.history.push.mockClear();
    instance.handleRedirect();
    expect(props.history.push).toHaveBeenCalledTimes(1);
    expect(props.history.push).toHaveBeenCalledWith('/test/redirect/after/login', { replace: true });
  });

  it('should do nothing if no redirect after login and not ready', () => {
    props.history.push.mockClear();
    instance.handleRedirect();
    expect(props.history.push).not.toHaveBeenCalled();
  });

  it('should redirect to summary report for reporting users', () => {
    wrapper.setProps({
      currentUser: {
        access_level: 'reporting'
      },
      ready: true
    });
    props.history.push.mockClear();
    instance.handleRedirect();
    expect(props.history.push).toHaveBeenCalledTimes(1);
    expect(props.history.push).toHaveBeenCalledWith('/reports/summary', { replace: true });
  });

  it('should redirect based on config', () => {
    wrapper.setProps({ ready: true });
    props.history.push.mockClear();
    instance.handleRedirect();
    expect(props.history.push).toHaveBeenCalledTimes(1);
    expect(props.history.push).toHaveBeenCalledWith('/config/splash', { replace: true });
  });

  it('should handle a redirect on mount', () => {
    jest.spyOn(instance, 'handleRedirect');
    instance.componentDidMount();
    expect(instance.handleRedirect).toHaveBeenCalledTimes(1);
  });

  it('should handle a redirect on update', () => {
    jest.spyOn(instance, 'handleRedirect');
    instance.componentDidUpdate();
    expect(instance.handleRedirect).toHaveBeenCalledTimes(1);
  });

});
