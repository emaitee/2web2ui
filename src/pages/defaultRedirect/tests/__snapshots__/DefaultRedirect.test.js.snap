// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DefaultRedirect should render correctly by default 1`] = `<Loading />`;
